File: src/test/java/com/alibaba/ttl/TtlCallableTest.java
package com.alibaba.ttl;

import com.alibaba.ttl.testmodel.Call;
import org.junit.AfterClass;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.*;

import static com.alibaba.ttl.Utils.CHILD;
import static com.alibaba.ttl.Utils.PARENT_AFTER_CREATE_TTL_TASK;
import static com.alibaba.ttl.Utils.PARENT_MODIFIED_IN_CHILD;
import static com.alibaba.ttl.Utils.PARENT_UNMODIFIED_IN_CHILD;
import static com.alibaba.ttl.Utils.assertTtlInstances;
import static com.alibaba.ttl.Utils.captured;
import static com.alibaba.ttl.Utils.createTestTtlValue;
import static com.alibaba.ttl.Utils.expandThreadPool;
import static org.hamcrest.CoreMatchers.containsString;
import static org.hamcrest.CoreMatchers.instanceOf;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.fail;


/**
 * @author Jerry Lee (oldratlee at gmail dot com)
 */
public class TtlCallableTest {
    private static ExecutorService executorService = Executors.newFixedThreadPool(3);

    static {
        expandThreadPool(executorService);
    }

    @AfterClass
    public static void afterClass() throws Exception {
        executorService.shutdown();
        executorService.awaitTermination(100, TimeUnit.MILLISECONDS);
        if (!executorService.isTerminated()) fail("Fail to shutdown thread pool");
    }

    @Test
    public void test_TtlCallable_inSameThread() throws Exception {
        ConcurrentMap<String, TransmittableThreadLocal<String>> ttlInstances = createTestTtlValue();

        Call call = new Call("1", ttlInstances);
        TtlCallable<String> ttlCallable = TtlCallable.get(call);

        // create after new Task, won't see parent value in in task!
        TransmittableThreadLocal<String> after = new TransmittableThreadLocal<>();
        after.set(PARENT_AFTER_CREATE_TTL_TASK);
        ttlInstances.put(PARENT_AFTER_CREATE_TTL_TASK, after);

        String ret = ttlCallable.call();
        assertEquals("ok", ret);

        // child Inheritable
        assertTtlInstances(call.captured,
                PARENT_UNMODIFIED_IN_CHILD, PARENT_UNMODIFIED_IN_CHILD,
                PARENT_MODIFIED_IN_CHILD + 1, PARENT_MODIFIED_IN_CHILD,
                CHILD + 1, CHILD + 1
        );

        // child do not effect parent
        assertTtlInstances(captured(ttlInstances),
                PARENT_UNMODIFIED_IN_CHILD, PARENT_UNMODIFIED_IN_CHILD,
                PARENT_MODIFIED_IN_CHILD, PARENT_MODIFIED_IN_CHILD, // restored after call!
                PARENT_AFTER_CREATE_TTL_TASK, PARENT_AFTER_CREATE_TTL_TASK
        );
    }

    @Test
    public void test_TtlCallable_asyncWithExecutorService() throws Exception {
        ConcurrentMap<String, TransmittableThreadLocal<String>> ttlInstances = createTestTtlValue();

        Call call = new Call("1", ttlInstances);
        TtlCallable<String> ttlCallable = TtlCallable.get(call);

        // create after new Task, won't see parent value in in task!
        TransmittableThreadLocal<String> after = new TransmittableThreadLocal<>();
        after.set(PARENT_AFTER_CREATE_TTL_TASK);
        ttlInstances.put(PARENT_AFTER_CREATE_TTL_TASK, after);

        Future<String> future = executorService.submit(ttlCallable);
        assertEquals("ok", future.get());

        // child Inheritable
        assertTtlInstances(call.captured,
                PARENT_UNMODIFIED_IN_CHILD, PARENT_UNMODIFIED_IN_CHILD,
                PARENT_MODIFIED_IN_CHILD + 1, PARENT_MODIFIED_IN_CHILD,
                CHILD + 1, CHILD + 1
        );

        // child do not effect parent
        assertTtlInstances(captured(ttlInstances),
                PARENT_UNMODIFIED_IN_CHILD, PARENT_UNMODIFIED_IN_CHILD,
                PARENT_MODIFIED_IN_CHILD, PARENT_MODIFIED_IN_CHILD,
                PARENT_AFTER_CREATE_TTL_TASK, PARENT_AFTER_CREATE_TTL_TASK
        );
    }

    @Test
    public void test_removeSameAsNotSet() throws Exception {
        ConcurrentMap<String, TransmittableThreadLocal<String>> ttlInstances = createTestTtlValue();
        ttlInstances.get(PARENT_UNMODIFIED_IN_CHILD).remove();

        Call call = new Call("1", ttlInstances);
        TtlCallable<String> ttlCallable = TtlCallable.get(call);

        // create after new Task, won't see parent value in in task!
        TransmittableThreadLocal<String> after = new TransmittableThreadLocal<>();
        after.set(PARENT_AFTER_CREATE_TTL_TASK);
        ttlInstances.put(PARENT_AFTER_CREATE_TTL_TASK, after);

        Future<String> future = executorService.submit(ttlCallable);
        assertEquals("ok", future.get());

        // child Inheritable
        assertTtlInstances(call.captured,
                PARENT_MODIFIED_IN_CHILD + 1, PARENT_MODIFIED_IN_CHILD,
                CHILD + 1, CHILD + 1
        );

        // child do not effect parent
        assertTtlInstances(captured(ttlInstances),
                PARENT_MODIFIED_IN_CHILD, PARENT_MODIFIED_IN_CHILD,
                PARENT_AFTER_CREATE_TTL_TASK, PARENT_AFTER_CREATE_TTL_TASK
        );
    }

    @Test
    public void test_releaseTtlValueReferenceAfterCall() throws Exception {
        ConcurrentMap<String, TransmittableThreadLocal<String>> ttlInstances = createTestTtlValue();

        Call call = new Call("1", ttlInstances);
        TtlCallable<String> ttlCallable = TtlCallable.get(call, true);
        assertSame(call, ttlCallable.getCallable());

        Future<String> future = executorService.submit(ttlCallable);
        assertEquals("ok", future.get());

        future = executorService.submit(ttlCallable);
        try {
            future.get();
            fail();
        } catch (ExecutionException expected) {
            assertThat(expected.getCause(), instanceOf(IllegalStateException.class));
            assertThat(expected.getMessage(), containsString("TTL value reference is released after call!"));
        }
    }

    @Test
    public void test_get_same() throws Exception {
        Call call = new Call("1", null);
        TtlCallable<String> ttlCallable = TtlCallable.get(call);
        assertSame(call, ttlCallable.getCallable());
    }

    @Test
    public void test_get_idempotent() throws Exception {
        TtlCallable<String> call = TtlCallable.get(new Call("1", null));
        try {
            TtlCallable.get(call);
            fail();
        } catch (IllegalStateException e) {
            assertThat(e.getMessage(), containsString("Already TtlCallable"));
        }
    }

    @Test
    public void test_get_nullInput() throws Exception {
        assertNull(TtlCallable.get(null));
    }

    @Test
    public void test_gets() throws Exception {
        Callable<String> call1 = new Call("1", null);
        Callable<String> call2 = new Call("1", null);
        Callable<String> call3 = new Call("1", null);

        @SuppressWarnings("unchecked")
        List<TtlCallable<String>> callList = TtlCallable.gets(
                Arrays.asList(call1, call2, null, call3));

        assertEquals(4, callList.size());
        assertThat(callList.get(0), instanceOf(TtlCallable.class));
        assertThat(callList.get(1), instanceOf(TtlCallable.class));
        assertNull(callList.get(2));
        assertThat(callList.get(3), instanceOf(TtlCallable.class));
    }
}

File: src/test/java/com/alibaba/ttl/TtlCallableTest.kt
package com.alibaba.ttl

import com.alibaba.*
import com.alibaba.ttl.testmodel.Call
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.AnnotationSpec
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.nulls.shouldBeNull
import io.kotest.matchers.string.shouldContain
import io.kotest.matchers.types.shouldBeInstanceOf
import org.junit.Assert.*
import java.util.concurrent.Callable
import java.util.concurrent.ExecutionException
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

/**
 * @author Jerry Lee (oldratlee at gmail dot com)
 */
class TtlCallableTest : AnnotationSpec() {
    @Test
    fun test_TtlCallable_runInCurrentThread() {
        val ttlInstances = createParentTtlInstances()

        val call = Call("1", ttlInstances)


        val ttlCallable = TtlCallable.get(call)!!

        // create after new Task, won't see parent value in in task!
        createParentTtlInstancesAfterCreateChild(ttlInstances)


        // run in the *current* thread
        assertEquals("ok", ttlCallable.call())


        // child Inheritable
        assertChildTtlValues("1", call.copied)

        // child do not effect parent
        assertParentTtlValues(copyTtlValues(ttlInstances))
    }

    @Test
    fun test_TtlCallable_asyncRunByExecutorService() {
        val ttlInstances = createParentTtlInstances()

        val call = Call("1", ttlInstances)
        val ttlCallable = if (noTtlAgentRun()) TtlCallable.get(call) else call

        if (noTtlAgentRun()) {
            // create after new Task, won't see parent value in in task!
            createParentTtlInstancesAfterCreateChild(ttlInstances)
        }
        val future = executorService.submit(ttlCallable)
        if (!noTtlAgentRun()) {
            // create after new Task, won't see parent value in in task!
            createParentTtlInstancesAfterCreateChild(ttlInstances)
        }

        assertEquals("ok", future.get())


        // child Inheritable
        assertChildTtlValues("1", call.copied)

        // child do not effect parent
        assertParentTtlValues(copyTtlValues(ttlInstances))
    }

    @Test
    fun test_remove_sameAsNotSet() {
        val ttlInstances = createParentTtlInstances()


        // add and remove !!
        newTtlInstanceAndPut("add and removed!", ttlInstances).remove()

        val call = Call("1", ttlInstances)
        val ttlCallable = if (noTtlAgentRun()) TtlCallable.get(call) else call


        if (noTtlAgentRun()) {
            // create after new Task, won't see parent value in in task!
            createParentTtlInstancesAfterCreateChild(ttlInstances)
        }
        val future = executorService.submit(ttlCallable)
        if (!noTtlAgentRun()) {
            // create after new Task, won't see parent value in in task!
            createParentTtlInstancesAfterCreateChild(ttlInstances)
        }

        assertEquals("ok", future.get())


        // child Inheritable
        assertChildTtlValues("1", call.copied)

        // child do not affect parent
        assertParentTtlValues(copyTtlValues(ttlInstances))
    }

    @Test
    fun test_releaseTtlValueReferenceAfterCall() {
        val ttlInstances = createParentTtlInstances()

        val call = Call("1", ttlInstances)
        val ttlCallable = TtlCallable.get(call, true)!!
        assertSame(call, ttlCallable.callable)

        assertEquals("ok", executorService.submit(ttlCallable).get())


        val exception = shouldThrow<ExecutionException> {
            executorService.submit(ttlCallable).get()
        }
        exception.cause.shouldBeInstanceOf<IllegalStateException>()
        exception.message shouldContain "TTL value reference is released after call!"
    }

    @Test
    fun test_get_same() {
        val call = Call("1")
        val ttlCallable = TtlCallable.get(call)!!
        assertSame(call, ttlCallable.callable)
    }

    @Test
    fun test_get_idempotent() {
        val call = TtlCallable.get(Call("1"))

        shouldThrow<java.lang.IllegalStateException> {
            TtlCallable.get(call)
        }.message shouldContain "Already TtlCallable"
    }

    @Test
    fun test_get_nullInput() {
        TtlCallable.get<Any>(null).shouldBeNull()
    }

    @Test
    fun test_gets() {
        val call1 = Call("1")
        val call2 = Call("2")
        val call3 = Call("3")

        val callList = TtlCallable.gets(
            listOf<Callable<String>?>(call1, call2, null, call3)
        )

        callList.shouldHaveSize(4)
        callList[0].shouldBeInstanceOf<TtlCallable<*>>()
        callList[1].shouldBeInstanceOf<TtlCallable<*>>()
        callList[2].shouldBeNull()
        callList[3].shouldBeInstanceOf<TtlCallable<*>>()
    }

    @Test
    fun test_unwrap() {
        assertNull(TtlCallable.unwrap<String>(null))

        val callable = Callable { "hello" }
        val ttlCallable = TtlCallable.get(callable)


        assertSame(callable, TtlCallable.unwrap(callable))
        assertSame(callable, TtlCallable.unwrap(ttlCallable))

        assertSame(callable, TtlUnwrap.unwrap(callable))
        assertSame(callable, TtlUnwrap.unwrap(ttlCallable))


        assertEquals(listOf(callable), TtlCallable.unwraps(listOf(callable)))
        assertEquals(listOf(callable), TtlCallable.unwraps(listOf(ttlCallable)))
        assertEquals(listOf(callable, callable), TtlCallable.unwraps(listOf(ttlCallable, callable)))
        assertEquals(listOf<Callable<String>>(), TtlCallable.unwraps<String>(null))
    }

    @AfterAll
    fun afterAll() {
        executorService.shutdown()
        assertTrue("Fail to shutdown thread pool", executorService.awaitTermination(1, TimeUnit.SECONDS))
    }

    companion object {
        private val executorService = Executors.newFixedThreadPool(3).also { expandThreadPool(it) }
    }
}


